File: todo-frontend/src/api/todos.ts
import type { Todo, TodoCreate, TodoUpdate } from './types';

// 1. Đọc Base URL từ file .env [cite: 56]
const API_URL = import.meta.env.VITE_API_URL;

// Hàm helper để xử lý chung các phản hồi API
async function handleApiResponse(response: Response) {
  if (!response.ok) {
    // Xử lý lỗi chuẩn (400, 404, 422, 500) [cite: 17, 64, 65]
    const errorData = await response.json().catch(() => ({}));
    const message = errorData.detail || `Error ${response.status}`;
    throw new Error(message);
  }
  // Nếu là DELETE (204) thì không cần parse JSON [cite: 15]
  if (response.status === 204) return;
  return response.json();
}

// 2. GET /todos [cite: 12]
export const getTodos = async (): Promise<Todo[]> => {
  const response = await fetch(`${API_URL}/todos`);
  return handleApiResponse(response);
};

// 3. POST /todos [cite: 13]
export const createTodo = async (data: TodoCreate): Promise<Todo> => {
  const response = await fetch(`${API_URL}/todos`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  return handleApiResponse(response);
};

// 4. PATCH /todos/{id} [cite: 14]
export const updateTodo = async (id: number, data: TodoUpdate): Promise<Todo> => {
  const response = await fetch(`${API_URL}/todos/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  return handleApiResponse(response);
};

// 5. DELETE /todos/{id} [cite: 15]
export const deleteTodo = async (id: number): Promise<void> => {
  const response = await fetch(`${API_URL}/todos/${id}`, {
    method: 'DELETE',
  });
  return handleApiResponse(response);
};

File: todo-backend/app/routers/todos.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from .. import models, schemas # Dùng .. để import từ thư mục cha
from ..db import get_db

router = APIRouter(
    prefix="/todos", # Thêm /todos vào trước tất cả các endpoint
    tags=["todos"]    # Nhóm các API này lại trong /docs
)

# GET /todos [cite: 12]
@router.get("/", response_model=List[schemas.Todo])
def get_all_todos(db: Session = Depends(get_db)):
    todos = db.query(models.Todo).all()
    return todos

# POST /todos [cite: 13]
@router.post("/", response_model=schemas.Todo, status_code=status.HTTP_201_CREATED)
def create_todo(todo: schemas.TodoCreate, db: Session = Depends(get_db)):
    # schema TodoCreate đã tự động validate (không rỗng, <= 140)
    # Nếu fail, FastAPI tự trả về lỗi 422 [cite: 17]
    new_todo = models.Todo(title=todo.title)
    db.add(new_todo)
    db.commit()
    db.refresh(new_todo)
    return new_todo

# Hàm helper để tìm todo (tránh lặp code)
def get_todo_by_id(id: int, db: Session):
    db_todo = db.query(models.Todo).filter(models.Todo.id == id).first()
    if db_todo is None:
        # Lỗi 404 nếu không tồn tại [cite: 17]
        raise HTTPException(status_code=404, detail="Todo not found")
    return db_todo

# PATCH /todos/{id} [cite: 14]
@router.patch("/{id}", response_model=schemas.Todo)
def update_todo(id: int, todo_update: schemas.TodoUpdate, db: Session = Depends(get_db)):
    db_todo = get_todo_by_id(id, db)
    
    # Lấy data đã gửi, chỉ lấy các trường được set (không null)
    update_data = todo_update.dict(exclude_unset=True)
    
    for key, value in update_data.items():
        setattr(db_todo, key, value)
    
    db.commit()
    db.refresh(db_todo)
    return db_todo

# DELETE /todos/{id} [cite: 15]
@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_todo(id: int, db: Session = Depends(get_db)):
    db_todo = get_todo_by_id(id, db)
    db.delete(db_todo)
    db.commit()
    return None # Trả về 204 No Content